src/3d/qgsskyboxentity.cpp

```diff
@@ -0,0 +1,193 @@
+  , mFilterKey( new Qt3DRender::QFilterKey( this ) )
+  , mGl3RenderPass( new Qt3DRender::QRenderPass( this ) )
+  , mMesh( new Qt3DExtras::QCuboidMesh( this ) )
+  , mGammaStrengthParameter( new Qt3DRender::QParameter( QStringLiteral( "gammaStrength" ), 0.0f ) )
+  , mTextureParameter( new Qt3DRender::QParameter( this ) )
+{
+  mGl3Technique->graphicsApiFilter()->setApi( Qt3DRender::QGraphicsApiFilter::OpenGL );
+  mGl3Technique->graphicsApiFilter()->setMajorVersion( 3 );
+  mGl3Technique->graphicsApiFilter()->setMinorVersion( 3 );
+  mGl3Technique->graphicsApiFilter()->setProfile( Qt3DRender::QGraphicsApiFilter::CoreProfile );
+
+  mFilterKey->setParent( mEffect );
+  mFilterKey->setName( QStringLiteral( "renderingStyle" ) );
+  mFilterKey->setValue( QStringLiteral( "forward" ) );
+
+  mGl3Technique->addFilterKey( mFilterKey );
+
+  Qt3DRender::QCullFace *cullFront = new Qt3DRender::QCullFace();
+  cullFront->setMode( Qt3DRender::QCullFace::Front );
+  Qt3DRender::QDepthTest *depthTest = new Qt3DRender::QDepthTest();
+  depthTest->setDepthFunction( Qt3DRender::QDepthTest::LessOrEqual );
+  Qt3DRender::QSeamlessCubemap *seamlessCubemap = new Qt3DRender::QSeamlessCubemap();
+
+  mGl3RenderPass->addRenderState( cullFront );
+  mGl3RenderPass->addRenderState( depthTest );
+  mGl3RenderPass->addRenderState( seamlessCubemap );
+
+  mGl3Technique->addRenderPass( mGl3RenderPass );
+
+  mEffect->addTechnique( mGl3Technique );
+
+  mMaterial->setEffect( mEffect );
+  mMaterial->addParameter( mGammaStrengthParameter );
+  mMaterial->addParameter( mTextureParameter );
+
+  mMesh->setXYMeshResolution( QSize( 2, 2 ) );
+  mMesh->setXZMeshResolution( QSize( 2, 2 ) );
+  mMesh->setYZMeshResolution( QSize( 2, 2 ) );
+
+//   TODO: change the kybox position according to camera
+  Qt3DCore::QTransform *transform = new Qt3DCore::QTransform( this );
+  transform->setTranslation( QVector3D( 0.0f, 0.0f, 0.0f ) );
+  transform->setScale( 1000.0f );
+  addComponent( transform );
+
+  addComponent( mMesh );
+  addComponent( mMaterial );
+}
+
+// HDR skybox
+
+QgsHDRSkyboxEntity::QgsHDRSkyboxEntity( const QString &hdrTexturePath, QNode *parent )
+  : QgsSkyboxEntity( parent )
+  , mHDRTexturePath( hdrTexturePath )
+  , mLoadedTexture( new Qt3DRender::QTextureLoader( parent ) )
+  , mGlShader( new Qt3DRender::QShaderProgram( this ) )
+{
+  mLoadedTexture->setGenerateMipMaps( false );
+  mGlShader->setVertexShaderCode( Qt3DRender::QShaderProgram::loadSource( QUrl( QStringLiteral( "qrc:/shaders/skybox.vert" ) ) ) );
+  mGlShader->setFragmentShaderCode( Qt3DRender::QShaderProgram::loadSource( QUrl( QStringLiteral( "qrc:/shaders/hdr_skybox.frag" ) ) ) );
+  mGl3RenderPass->setShaderProgram( mGlShader );
+
+  mTextureParameter->setName( "skyboxTexture" );
+  mTextureParameter->setValue( QVariant::fromValue( mLoadedTexture ) );
+
+  reloadTexture();
+}
+
+void QgsHDRSkyboxEntity::reloadTexture()
+{
+  mLoadedTexture->setSource( QUrl::fromUserInput( mHDRTexturePath ) );
```

Comment:
This will need to use QgsImageCache, just like the textures in the phong material (otherwise embedded textures won't work)

Reply:
I tried to do this for HDR images but apparently that's not supported with QImage.